File: 0202-happy-number/0202-happy-number.js
/**
 * Determines if a number is a happy number.
 *
 * @param {number} n
 * @return {boolean}
 */
function isHappy(n) {
    /**
     * 각 자리 숫자의 제곱 합을 계산
     *
     * 예: 19 → 1² + 9² = 82
     *
     * @param {number} num
     * @return {number}
     */
    const getNext = (num) => {
        let sumOfSquares = 0;

        while (num !== 0) {
            const digit = num % 10;

            sumOfSquares += digit ** 2;

            num = Math.floor(num / 10);
        }

        return sumOfSquares;
    };

    // 이미 방문한 숫자를 저장하여 사이클 감지
    const visitedNumbers = new Set();

    // 1이 될 때까지 반복
    while (n !== 1) {
        const nextNumber = getNext(n);

        // 이미 방문한 숫자라면 사이클 발생
        if (visitedNumbers.has(nextNumber)) {
            return false;
        }

        visitedNumbers.add(nextNumber);

        n = nextNumber;
    }

    return true;
}

File: 0202-happy-number/0202-happy-number.java
class Solution {
    /**
     * Determines if a number is a "happy number".
     * A happy number is defined by the following process:
     * - Starting with any positive integer, replace the number by the sum of the squares of its digits
     * - Repeat the process until the number equals 1 (happy) or loops endlessly in a cycle (not happy)
     * 
     * @param n The positive integer to check
     * @return true if n is a happy number, false otherwise
     */
    public boolean isHappy(int n) {
        // Set to track visited numbers and detect cycles
        Set<Integer> visitedNumbers = new HashSet<>();
      
        // Continue until we reach 1 (happy) or find a cycle (not happy)
        while (n != 1 && !visitedNumbers.contains(n)) {
            // Mark current number as visited
            visitedNumbers.add(n);
          
            // Calculate sum of squares of digits
            int sumOfSquares = 0;
            while (n != 0) {
                int digit = n % 10;  // Extract last digit
                sumOfSquares += digit * digit;  // Add square of digit to sum
                n /= 10;  // Remove last digit
            }
          
            // Update n with the new sum for next iteration
            n = sumOfSquares;
        }
      
        // If we exited because n equals 1, it's happy; otherwise, we found a cycle
        return n == 1;
    }
}
